cmds/report.c

```diff
@@ -20,6 +21,213 @@ enum {
 /* maximum length of symbol */
 static int maxlen = 20;
 
+static LIST_HEAD(output_fields);
+
+static void print_total_time(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->total.sum;
+
+	print_time_unit(d);
+}
+
+static void print_total_time_avg(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->total.avg;
+
+	print_time_unit(d);
+}
+
+static void print_total_time_min(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->total.min;
+
+	print_time_unit(d);
+}
+
+static void print_total_time_max(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->total.max;
+
+	print_time_unit(d);
+}
+
+static void print_self_time(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->self.sum;
+
+	print_time_unit(d);
+}
+
+static void print_self_time_avg(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->self.avg;
+
+	print_time_unit(d);
+}
+
+static void print_self_time_min(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->self.min;
+
+	print_time_unit(d);
+}
+
+static void print_self_time_max(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	uint64_t d;
+
+	d = node->self.max;
+
+	print_time_unit(d);
+}
+
+static void print_call(struct field_data *fd)
+{
+	struct uftrace_report_node *node = fd->arg;
+	pr_out("%10"PRIu64 "", node->call);
+}
+
+static struct display_field field_total_time = {
+	.id      = REPORT_F_TOTAL_TIME,
+	.name    = "total-time",
```

Comment:
Could you give me any opinions about composition of field and field name?
at this time, I just tried to follow naming convention and architecture of previous codes.

Reply:
I think this is ok.

Reply:
OK, I took a look at this again and I think we should not do this.  By using a macro generates the functions and struct we don't need to care about the commonality and can get rid of this.

Reply:
Thank you for your comment.
Do you mean macro that generates functions and "display_field" struct?
In fact, I'd thought about a macro, but the number of common fields is not much compared to SORT_KEY.
so I hope to discuss the direction of macro briefly.
These are just psuedo code to assume the shape.

1. base case (not using --diff)
1.1 define macro for base case 
print function is different as option, so I guess macro has to be different as option. 
1.1.1 macro for struct.
macro to generate struct can be used commonly if print function is divided as options. 
```
#define MACRO_STRUCT(_id, _name, _field, _header, _length, _opt)                                                \
static struct display_field field_##_name_##_opt = { \ //different as struct instance (1)
        .id      = _id,                              \ //different as field id(2)
        .name    = #_name,                           \ //different as field id(2)
        .header  = _header,                          \ //different as field contents (3)
        .length  = _length,                          \ //different as field contents (3)
        .print   = print_##_name_##_opt,             \ //different as field contents (3)
        .list    = LIST_HEAD_INIT(field_##_name_##_opt.list), \ // (1)
};
```
1.1.2  this macro to generate "time" print function and struct for base case.
```
#define MACRO_TIME(_id, _name, _field, _header, _length, _opt)  \
static void print_##_name(struct field_data *fd)    \
{ \
        struct uftrace_report_node *node = fd->arg; \
        print_time_unit(node->_field);\
} \
#define MACRO_STRUCT(_id, _name, _field, _header, _length, _opt)\
```
1.1.3 this macro to generate "call" print function and struct for base case.
```
#define MACRO_CALL(_id, _name, _field, _header, _length_, _opt) \
static void print_##_name(struct field_data *fd)    \ // call print function is not common with time print function.
{ \
        struct uftrace_report_node *node = fd->arg; \
        pr_out("%10"PRIu64 "", node->call); \
} \
#define MACRO_STRUCT(_id, _name, _field, _header, _length, _opt)\
```
1.2  using macro for base case (not using --diff)
maybe we can reduce some field by generating it in macro with the value of other argument.
simplicity and flexibility can't go together sometimes.
```
// default
MACRO_TIME(REPORT_F_TOTAL_TIME, total, total, "Total", 10, "");
MACRO_TIME(REPORT_F_TOTAL_TIME_AVG, total-avg, total.avg, "Total avg", 10, "");
MACRO_TIME(REPORT_F_TOTAL_TIME_MIN, total-min, total.min, "Total min", 10, "");
MACRO_TIME(REPORT_F_TOTAL_TIME_MAX, total-max, total.max, "Total max", 10, "");
MACRO_TIME(REPORT_F_SELF_TIME, self, self, "Self", 10, "");
MACRO_TIME(REPORT_F_SELF_TIME_AVG, self-avg, self.avg, "Self avg", 10, "");
MACRO_TIME(REPORT_F_SELF_TIME_MIN, self-min, self.min, "Self min", 10, "");
MACRO_TIME(REPORT_F_SELF_TIME_MAX, self-max, self-max, "Self max", 10, "");
MACRO_CALL(REPORT_F_CALL, call, call, "Call", 10, "");
```

2. diff case 
2.1 define macro 
define new one with print function for diff case
2.2  using new macro
```
MACRO(REPORT_F_TOTAL_TIME, total, total, "Total", 11, "diff");
MACRO(REPORT_F_TOTAL_TIME_AVG, total-avg, total.avg, "Total avg", 11, "diff");
MACRO(REPORT_F_TOTAL_TIME_MIN, total-min, total.min, "Total min", 11, "diff");
MACRO(REPORT_F_TOTAL_TIME_MAX, total-max, total.max, "Total max", 11, "diff");
MACRO(REPORT_F_SELF_TIME, self, self, "Self", 11, "diff");
MACRO(REPORT_F_SELF_TIME_AVG, self-avg, self.avg, "Self avg", 11, "diff");
MACRO(REPORT_F_SELF_TIME_MIN, self-min, self.min, "Self min", 11, "diff");
MACRO(REPORT_F_SELF_TIME_MAX, self-max, self-max, "Self max", 11, "diff");
MACRO_CALL(REPORT_F_CALL, call, call, "Call", 11, "diff");
```

3. diff full case 
3.1 define macro 
define new one with print function for diff full case.
3.2  using new macro
```
// diff : different macro with above because print function
MACRO(REPORT_F_TOTAL_TIME, total, total, "Total (diff)", 35, "diff_full");
MACRO(REPORT_F_TOTAL_TIME_AVG, total-avg, total.avg, "Total avg (diff)", 35, "diff_full");
MACRO(REPORT_F_TOTAL_TIME_MIN, total-min, total.min, "Total min (diff)", 35, "diff_full");
MACRO(REPORT_F_TOTAL_TIME_MAX, total-max, total.max, "Total max (diff)", 35, "diff_full");
MACRO(REPORT_F_SELF_TIME, self, self, "Self (diff)", 35, "diff_full");
MACRO(REPORT_F_SELF_TIME_AVG, self-avg, self.avg, "Self avg (diff)", 35, "diff_full");
MACRO(REPORT_F_SELF_TIME_MIN, self-min, self.min, "Self min (diff)", 35, "diff_full");
MACRO_CALL(REPORT_F_CALL, call, call, "Call", 11, "diff_full");
```

3. diff full percent case 
3.1 define macro 
define new one with print function for diff full percent case.
3.2  using new macro
```
// diff full percent: different macro with above because print function
MACRO(REPORT_F_TOTAL_TIME, total, total, "Total (diff)", xx, "diff_full_percent");
MACRO(REPORT_F_TOTAL_TIME_AVG, total-avg, total.avg, "Total avg (diff)", xx, "diff_full_percent");
MACRO(REPORT_F_TOTAL_TIME_MIN, total-min, total.min, "Total min (diff)", xx, "diff_full_percent");
MACRO(REPORT_F_TOTAL_TIME_MAX, total-max, total.max, "Total max (diff)", xx, "diff_full_percent");
MACRO(REPORT_F_SELF_TIME, self, self, "Self (diff)", xx, "diff_full_percent");
MACRO(REPORT_F_SELF_TIME_AVG, self-avg, self.avg, "Self avg (diff)", xx, "diff_full_percent");
MACRO(REPORT_F_SELF_TIME_MIN, self-min, self.min, "Self min (diff)", xx, "diff_full_percent";
MACRO(REPORT_F_SELF_TIME_MAX, self-max, self-max, "Self max (diff)", xx, "diff_full_percent");
MACRO_CALL(REPORT_F_CALL, call, call, "Call", 11, "diff_full_percent");
```
could you give me any opinion?

Reply:
my last comment attached here.
could please you check this?

Reply:
Sorry, I missed this.  So yeah, I like to have a macro generates both function and struct for each item.  The basic approach looks good, please go ahead!